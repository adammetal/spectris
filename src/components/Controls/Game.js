import { Button, FlexCenter, Input } from "../ui-kit";
import { useEffect, useState } from "react";

const Game = ({ id, url, onRemove, onAdd, onUpdate, onSpectate }) => {
  const [value, setValue] = useState(url);
  const [changed, setChanged] = useState(false);

  useEffect(() => {
    if (value !== url) {
      setChanged(true);
    } else {
      setChanged(false);
    }
  }, [value, url]);

  const handleChange = (e) => {
    setValue(e.target.value);
  };

  const handleRemove = () => {
    onRemove(id);
  };

  const handleSpectate = () => {
    onSpectate(id);
  };

  const handleAdd = () => {
    onAdd(value);
    setValue("");
  };

  const handleUpdate = () => {
    onUpdate(id, value);
  };

  const hasId = id !== null;

  return (
    <FlexCenter row>
      <Input
        id="source"
        placeholder="Game url"
        value={value}
        onChange={handleChange}
      />
      <Button disabled={!changed} onClick={hasId ? handleUpdate : handleAdd}>
        {hasId ? "Update" : "Add"}
      </Button>
      <Button onClick={handleSpectate} disabled={!hasId}>
        Spectate
      </Button>
      <Button onClick={handleRemove} disabled={!hasId}>
        Remove
      </Button>
    </FlexCenter>
  );
};

Game.defaultProps = {
  id: null,
  url: "",
  onRemove: () => {},
  onAdd: () => {},
  onUpdate: () => {},
  onSpectate: () => {},
};

export default Game;