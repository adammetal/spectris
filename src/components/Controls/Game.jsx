import { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Button, FlexCenter, Input } from '../ui-kit';

function Game({ id, url, onRemove, onAdd, onUpdate, onSpectate }) {
  const [value, setValue] = useState(url);
  const [changed, setChanged] = useState(false);

  useEffect(() => {
    if (value !== url) {
      setChanged(true);
    } else {
      setChanged(false);
    }
  }, [value, url]);

  const handleChange = (e) => {
    setValue(e.target.value);
  };

  const handleRemove = () => {
    onRemove(id);
  };

  const handleSpectate = () => {
    onSpectate(url);
  };

  const handleAdd = () => {
    onAdd(value);
    setValue('');
  };

  const handleUpdate = () => {
    onUpdate(id, value);
  };

  const hasId = id !== null;

  return (
    <FlexCenter row>
      <Input id="source" placeholder="Game url" value={value} onChange={handleChange} />
      <Button disabled={!changed} onClick={hasId ? handleUpdate : handleAdd}>
        {hasId ? 'Update' : 'Add'}
      </Button>
      <Button onClick={handleSpectate} disabled={!hasId}>
        Spectate
      </Button>
      <Button onClick={handleRemove} disabled={!hasId}>
        Remove
      </Button>
    </FlexCenter>
  );
}

Game.defaultProps = {
  id: null,
  url: '',
  onRemove: () => {},
  onUpdate: () => {},
  onSpectate: () => {},
  onAdd: () => {},
};

Game.propTypes = {
  id: PropTypes.string,
  url: PropTypes.string,
  onRemove: PropTypes.func,
  onUpdate: PropTypes.func,
  onSpectate: PropTypes.func,
  onAdd: PropTypes.func,
};

export default Game;
