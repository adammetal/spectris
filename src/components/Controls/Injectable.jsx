import PropTypes from 'prop-types';
import { Button } from '../ui-kit';

function Injectable({ file, onRemove }) {
  const name = file?.split('/')?.at(-1);

  return (
    <div className="bg-blue-600 text-yellow-50 p-2 flex justify-between items-baseline gap-2">
      <div className="text-[1rem]">{name}</div>
      <Button w="w-12" p="p-auto" onClick={() => onRemove(file)}>
        âœ–
      </Button>
    </div>
  );
}

Injectable.propTypes = {
  file: PropTypes.string.isRequired,
  onRemove: PropTypes.func.isRequired,
};

export default Injectable;
