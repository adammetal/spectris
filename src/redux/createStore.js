import { configureStore } from '@reduxjs/toolkit';
import matchesReducer from './features/matches';

const storage = (key) => ({
  get: () => {
    const raw = localStorage.getItem(key) ?? '{}';
    return JSON.parse(raw);
  },
  set: (value) => {
    const raw = JSON.stringify(value);
    localStorage.setItem(key, raw);
  },
});

const reduxState = storage('reduxState');

export default () => {
  const preloadedState = reduxState.get();

  const store = configureStore({
    reducer: {
      matches: matchesReducer,
    },
    preloadedState,
  });

  store.subscribe(() => {
    reduxState.set(store.getState());
  });

  return store;
};
