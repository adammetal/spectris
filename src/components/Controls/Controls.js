import { useCallback } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Button, FlexCenter } from "../ui-kit";
import Game from "./Game";
import {
  addMatch,
  removeMatchById,
  updateMatchById,
} from "../../redux/features/matches";

const Controls = () => {
  const dispatch = useDispatch();
  const urls = useSelector((state) => state.matches.urls);

  const handleAdd = useCallback(
    (url) => {
      dispatch(addMatch({ url }));
    },
    [dispatch, addMatch]
  );

  const handleRemove = useCallback(
    (id) => {
      dispatch(removeMatchById({ id }));
    },
    [dispatch, removeMatchById]
  );

  const handleUpdate = useCallback(
    (id, url) => {
      dispatch(updateMatchById({ id, url }));
    },
    [dispatch, updateMatchById]
  );

  return (
    <FlexCenter>
      {urls.map((game) => {
        return (
          <Game
            key={game.id}
            id={game.id}
            url={game.url}
            onRemove={handleRemove}
            onUpdate={handleUpdate}
          />
        );
      })}
      <Game onAdd={handleAdd} />
      <FlexCenter row>
        <Button>Clear All</Button>
      </FlexCenter>
    </FlexCenter>
  );
};

export default Controls;
