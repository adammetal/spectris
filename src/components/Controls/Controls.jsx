import { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { FlexCenter } from '../ui-kit';
import Game from './Game';
import { addMatch, removeMatchById, updateMatchById } from '../../redux/features/matches';

function Controls() {
  const dispatch = useDispatch();
  const urls = useSelector((state) => state.matches.urls);

  const handleAdd = useCallback(
    (url) => {
      dispatch(addMatch({ url }));
    },
    [dispatch, addMatch],
  );

  const handleRemove = useCallback(
    (id) => {
      dispatch(removeMatchById({ id }));
    },
    [dispatch, removeMatchById],
  );

  const handleUpdate = useCallback(
    (id, url) => {
      dispatch(updateMatchById({ id, url }));
    },
    [dispatch, updateMatchById],
  );

  const handleSpectate = useCallback((url) => {
    window.controls.createWindow(url);
  });

  return (
    <FlexCenter>
      {urls.map((game) => (
        <Game
          key={game.id}
          id={game.id}
          url={game.url}
          onRemove={handleRemove}
          onUpdate={handleUpdate}
          onSpectate={handleSpectate}
        />
      ))}
      <Game onAdd={handleAdd} />
    </FlexCenter>
  );
}

export default Controls;
