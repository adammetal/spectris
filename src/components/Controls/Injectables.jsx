import PropTypes from 'prop-types';
import { Panel, Button } from '../ui-kit';
import Injectable from './Injectable';

function Injectables({ files, title, hint, onSelectNew }) {
  return (
    <Panel title={title}>
      <p className="mb-2 p-2 bg-blue-400">{hint}</p>
      {files.map((f) => (
        <Injectable file={f} key={f} />
      ))}
      <div className="mt-4">
        <Button onClick={onSelectNew} w="w-auto">
          Add new css file
        </Button>
      </div>
    </Panel>
  );
}

Injectables.propTypes = {
  files: PropTypes.arrayOf(PropTypes.string).isRequired,
  title: PropTypes.string.isRequired,
  hint: PropTypes.string.isRequired,
  onSelectNew: PropTypes.func.isRequired,
};

export default Injectables;
