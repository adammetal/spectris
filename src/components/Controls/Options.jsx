import { useDispatch, useSelector } from 'react-redux';
import Injectables from './Injectables';
import { addCss, addJs } from '../../redux/features/injector';

function Options() {
  const { css, js } = useSelector((state) => state.injector);
  const dispatch = useDispatch();

  const handleSelectCssFile = async () => {
    const { filePaths } = await window.spectate.selectInjectableCss();
    if (filePaths && filePaths.length) {
      dispatch(addCss({ file: filePaths[0] }));
    }
  };

  const handleSelectJsFile = async () => {
    const { filePaths } = await window.spectate.selectInjectableJs();
    if (filePaths && filePaths.length) {
      dispatch(addJs({ file: filePaths[0] }));
    }
  };

  return (
    <div className="flex gap-4 w-8/12">
      <Injectables
        files={css}
        title="Extra Css styles"
        onSelectNew={handleSelectCssFile}
        hint="When the inspector opens, you can inject extra css files. They will applied to the inspected game."
      />
      <Injectables
        files={js}
        title="Extra JavaScript codes"
        onSelectNew={handleSelectJsFile}
        hint="When the inspector opens, you can inject extra javascript files. The scripts will start runnig shortly after the inspector loads the game."
      />
    </div>
  );
}

export default Options;
