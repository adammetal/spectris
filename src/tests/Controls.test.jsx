import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import createStore from '../redux/createStore';
import Controls from '../components/Controls';

describe('Controls test', () => {
  let container;
  let store;

  beforeEach(() => {
    window.localStorage.setItem('reduxState', '{}');

    store = createStore();

    const result = render(
      <Provider store={store}>
        <Controls />
      </Provider>,
    );

    container = result.container;
  });

  test('snapshot', () => {
    expect(container).toMatchSnapshot();
  });

  test('default state', () => {
    const add = screen.getByText('Add');
    const spectate = screen.getByText('Spectate');
    const remove = screen.getByText('Remove');

    expect(add).toBeDisabled();
    expect(spectate).toBeDisabled();
    expect(remove).toBeDisabled();
  });

  test('type url', () => {
    const input = screen.getByPlaceholderText('Game url');
    const add = screen.getByText('Add');
    const spectate = screen.getByText('Spectate');
    const remove = screen.getByText('Remove');

    fireEvent.change(input, { target: { value: 'test' } });

    expect(add).toBeEnabled();
    expect(spectate).toBeDisabled();
    expect(remove).toBeDisabled();
  });

  test('Add url', async () => {
    const input = screen.getByPlaceholderText('Game url');
    const add = screen.getByText('Add');

    fireEvent.change(input, { target: { value: 'test' } });
    fireEvent.click(add);

    const update = await screen.findByText('Update');
    const newUrl = await screen.findByDisplayValue('test');

    expect(update).toBeDisabled();
    fireEvent.change(newUrl, { target: { value: 'updated' } });

    expect(update).toBeEnabled();
    fireEvent.click(update);

    expect(update).toBeDisabled();
  });

  test('Remove url', async () => {
    const input = screen.getByPlaceholderText('Game url');
    const add = screen.getByText('Add');

    fireEvent.change(input, { target: { value: 'test' } });
    fireEvent.click(add);

    const [remove] = await screen.findAllByText('Remove');
    fireEvent.click(remove);

    const inputs = screen.getAllByPlaceholderText('Game url');
    expect(inputs).toHaveLength(1);
  });
});
