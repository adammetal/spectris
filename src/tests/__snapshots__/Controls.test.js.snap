// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls test snapshot 1`] = `
<div>
  <div
    class="flex flex-col justify-center items-center gap-2 p-2"
  >
    <div
      class="flex flex-row justify-center items-center gap-2 p-2"
    >
      <input
        class="outline-none p-2 border-2 border-cyan-600 w-80"
        id="source"
        placeholder="Game url"
        value=""
      />
      <button
        class="border-2 w-24 p-2 border-cyan-600 opacity-50"
        disabled=""
      >
        Add
      </button>
      <button
        class="border-2 w-24 p-2 border-cyan-600 opacity-50"
        disabled=""
      >
        Spectate
      </button>
      <button
        class="border-2 w-24 p-2 border-cyan-600 opacity-50"
        disabled=""
      >
        Remove
      </button>
    </div>
    <div
      class="flex flex-row justify-center items-center gap-2 p-2"
    >
      <button
        class="border-2 w-24 p-2 border-cyan-600 "
      >
        Clear All
      </button>
    </div>
  </div>
</div>
`;
